### PR TITLE
Fix to "Shallow git clone (#18491)" for external PRs

### PR DESCRIPTION
**Pull request by brad-decker**
Fixes to "Shallow git clone (#18491)" for external PRs

## Explanation
#18491 -- see the discussion after merge, this is not working for all forks.

### CODE REVIEW

**Review by Gudahtt — DISMISSED** [review]
LGTM!

Needs an update after #18784 is merged to get a passing CI check

File: .circleci/config.yml [hunk 1]
```diff
@@ -40,7 +40,23 @@ aliases:
       mkdir -p ~/.ssh
       echo github.com ssh-ed25519 $GITHUB_SSH_KEY >> ~/.ssh/known_hosts
 
-      git clone --depth 1 "$CIRCLE_REPOSITORY_URL" --branch "$CIRCLE_BRANCH" .
+      # Clone remote and check the commit ID of the checked out code (sh version of bash `elif [[ "$CIRCLE_BRANCH" =~ ^pull\/* ]]`)
```
Comment (Gudahtt) [inline .circleci/config.yml:43]: What is this comment trying to say? At first I thought it was talking about the following line, but it seems unrelated.

It's also odd to see something described as the "sh version" of a bash snippet, given that this _is_ a Bash script.

File: .circleci/config.yml [hunk 2]
```diff
@@ -40,7 +40,23 @@ aliases:
...
+      if [ -n "${CIRCLE_TAG-}" ]; then
+        # tag
+        git clone --depth 1 "$CIRCLE_REPOSITORY_URL" .
+        git fetch --depth 1 --force origin "+refs/tags/${CIRCLE_TAG}:refs/tags/${CIRCLE_TAG}"
+        git reset --hard "$CIRCLE_SHA1"
+        git checkout -q "$CIRCLE_TAG"
+      elif [[ $(echo $CIRCLE_BRANCH | grep -E ^pull\/[0-9]+$) ]]; then
+        # pull request
+        git clone --depth 1 "$CIRCLE_REPOSITORY_URL" .
+        git fetch --depth 1 --force origin "${CIRCLE_BRANCH}/head:remotes/origin/${CIRCLE_BRANCH}"
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply .circleci/config.yml:53]: @HowardBraham

File: .circleci/config.yml (hunk 1 above)
Comment (HowardBraham) [reply .circleci/config.yml]: That comment was coming from here, but yeah I have revised it now
https://github.com/guitarrapc/git-shallow-clone-orb/blob/main/src/commands/checkout.yml

File: .circleci/config.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (HowardBraham) [reply .circleci/config.yml:53]: I didn't completely follow what the `git fetch` step is doing either, but this is what the standard CircleCI `checkout` command does for the case of an external PR (specifically pull/18748):

```
git clone --no-checkout "$CIRCLE_REPOSITORY_URL" .
git fetch --force origin +refs/pull/18748/head:refs/remotes/origin/pull/18748
git checkout --force -B "$CIRCLE_BRANCH" "$CIRCLE_SHA1"
git --no-pager log --no-color -n 1 --format='HEAD is now at %h %s'
```

**Review by legobeat — DISMISSED** [review]

**Comment by HowardBraham** [comment]
@legobeat @brad-decker @Gudahtt It's been a little while, but I'm coming back to this PR.

I have tested it with a forked PR #19573 and it passes all the tests.

I also tried to do it with a PR from a tag, but the "Create pull request" button disappears, as seen below:

image

File: .circleci/config.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (HowardBraham) [reply .circleci/config.yml]: @Gudahtt @legobeat I'm having trouble finding a single-step clone command that actually does the correct thing in the case of a PR from an external repo.  The following do not work:

```
git clone --depth 1 -b "$CIRCLE_BRANCH" "$CIRCLE_REPOSITORY_URL" .
git clone --depth 1 -b pull/19683/ https://github.com/MetaMask/metamask-extension .
```

```
git clone --depth 1 "$CIRCLE_REPOSITORY_URL" --branch "$CIRCLE_BRANCH" --origin "${CIRCLE_BRANCH}/head:remotes/origin/${CIRCLE_BRANCH}" .
git clone --depth 1 https://github.com/MetaMask/metamask-extension --branch pull/19683/ --origin pull/19693/head:remotes/origin/pull/19693
```
The only thing I can get working is that 3-step clone, fetch, checkout.  Do you guys have any ideas?

**Review by tuanggo — COMMENTED** [review]
ok

File: .circleci/config.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply .circleci/config.yml]: Hmm, I am not sure. Thoughts on bringing it back to the way it was for now, and leaving a comment referencing the source? Then we can improve it later.

**Comment by brad-decker** [comment]
@HowardBraham I think you should pull your changes into a new branch and open a new PR so that we can view this with fresh eyes. You may reference this PR in your description for the background.

File: .circleci/config.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (HowardBraham) [reply .circleci/config.yml:70]: Okay done, it's back to the way it was, with a comment about the source
